### PR TITLE
Add more endpoints and basic auth implementation

### PR DESCRIPTION
**Pull request by johnnymcmike**
This PR adds a handful of API endpoints:
-api/events/all
-api/events/add
-api/auth/login
-api/auth/logout
-api/applicants/count
-api/applicants/count/bymonth
-api/applicants/[applicantKey]
-api/applicants/[applicantKey]/delete

And an entirely reworked, functioning auth implementation based on Firebase Auth and this [next-firebase-auth](https://github.com/gladly-team/next-firebase-auth) package, rather than the old strategy of trying to leverage nextauth using firebase as a DB.

Notes for reviewers:
-Currently there is nothing role-based in this implementation, all of the new & old endpoints (save for api/events/all, since this is meant to be public) have a check to see if the user is signed in and return a 401 if they are not, letting them through otherwise. This is currently done with a copy-pasted if block, which will later be replaced by a function that instead checks if their account has the right roles to do that thing. TL;DR: Everyone who makes an account is basically an admin, don't merge this into main yet.

-Login page is going to have to tweak its implementation somewhat to work with this new package. logintest.js and authtest.js have been added as example sign-in and "are you signed in" pages, to be removed when appropriate. Currently logging out can only be done by navigating to /api/auth/logout.

### CODE REVIEW

File: src/pages/api/applicants/count/bymonth.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { getUserFromCookies } from 'next-firebase-auth';
```
Comment (Sanjay-B) [inline src/pages/api/applicants/count/bymonth.js:1]: What is the purpose of this endpoint?

File: src/pages/api/applicants/count/index.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { Firestore_Database } from '../../../../database';
+import { getUserFromCookies } from 'next-firebase-auth';
+
+//endpoint that returns the number of applicant documents in the applicants collection of the database
```
Comment (Sanjay-B) [inline src/pages/api/applicants/count/index.js:4]: Could you not you just get the length of getall in a previous endpoint? Is this different?

File: src/pages/api/applicants/index.js [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
+import { Firestore_Database } from '../../../database';
 
 export default async function handler(req, res) {
-    if(req.method !== 'GET')
-    {
-      res.status(405).send({message: 'Only GET requests at this endpoint, please'})
-      return;
-    }
-    res.status(200).json(await getAllApplicants());
+  console.log(await getUserFromCookies({ req: req }));
```
Comment (Sanjay-B) [inline src/pages/api/applicants/index.js:5]: Nitpick: We don't need this to clutter our logs in prod.

File: src/pages/api/applicants/submit.js [hunk 4]
```diff
@@ -1,15 +1,33 @@
+import { Firestore_Database } from '../../../database';
+import { getUserFromCookies } from 'next-firebase-auth';
+
+
+export default async function handler(req, res) {
```
Comment (Sanjay-B) [inline src/pages/api/applicants/submit.js:5]: Don't need a handler. It's only for GET requests.

File: src/pages/api/auth/login.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { setAuthCookies } from 'next-firebase-auth'
+import initAuth from "../../../initAuth";
+initAuth()
+
+const handler = async (req, res) => {
```
Comment (Sanjay-B) [inline src/pages/api/auth/login.js:5]: Is this a login or really, just a refresh?

File: src/pages/api/events/add.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { Firestore_Database } from '../../../database';
+import { getUserFromCookies } from 'next-firebase-auth';
+
+
+export default async function handler(req, res) {
```
Comment (Sanjay-B) [inline src/pages/api/events/add.js:5]: Don't need handler. Only GET requests.

File: src/pages/authtest.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import React from 'react'
```
Comment (Sanjay-B) [inline src/pages/authtest.js:1]: You should make this an actual unit test. Because, if this goes in prod, it'll act as an actual page.

File: src/pages/logintest.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+//example component from
```
Comment (Sanjay-B) [inline src/pages/logintest.js:1]: You should make this a Unit Test & not a page.

**Review by Sanjay-B — CHANGES_REQUESTED** [review]
Impressive stuff. Requested changes / clarification.

File: src/pages/api/applicants/count/bymonth.js (hunk 1 above)
Comment (johnnymcmike) [reply src/pages/api/applicants/count/bymonth.js:1]: Metric dashboard screenshots that Aimi showed me had a histogram with new applications per-month. This is supposed to accommodate for that.

File: src/pages/api/applicants/index.js (hunk 3 above)
Comment (johnnymcmike) [reply src/pages/api/applicants/index.js:5]: Thought I already removed this. Done

File: src/pages/authtest.js (hunk 7 above)
Comment (johnnymcmike) [reply src/pages/authtest.js:1]: It's intentionally acting as an actual page right now, not intended for prod, meant to be removed whenever we don't need it. I copied it from their example app just for the sake of ease of testing functionality. I don't know how to write actual unit tests, but I remember you saying something about Automation putting out some kind of guide, so I'll look over that.

File: src/pages/api/auth/login.js (hunk 5 above)
Comment (johnnymcmike) [reply src/pages/api/auth/login.js:5]: Not sure. It was in their example app, and it seems to be required in some way because the path to this endpoint is a required paramater for initializing next-firebase-auth. The actual login functionality does seem to be at least partially elsewhere, look at the new component that logintest.js uses if you want to know more

Comment (Sanjay-B) [reply src/pages/api/auth/login.js:5]: Interesting. Thanks for the clarification.

File: src/pages/authtest.js (hunk 7 above)
Comment (Sanjay-B) [reply src/pages/authtest.js:1]: Follow up with Vincent & Isaac about it. They should know. Resolving this.

File: src/pages/api/applicants/count/index.js (hunk 2 above)
Comment (johnnymcmike) [reply src/pages/api/applicants/count/index.js:4]: Yeah, but where would I be returning that? I can just delete this file if it isn't really useful to us to have a total count of all applications ever as an endpoint. Or alternatively just move that count function to the file the get all endpoint is in

Comment (Sanjay-B) [reply src/pages/api/applicants/count/index.js:4]: You could make a separate method and just, get the length of the return. In this case, ``getAll().keys()``.

File: src/pages/api/applicants/count/bymonth.js (hunk 1 above)
Comment (Sanjay-B) [reply src/pages/api/applicants/count/bymonth.js:1]: Oh, so this is filtering. I'd make a search method that allows you to filter stuff out instead of this. i.e limit, cursor, sortOrder, etc.

File: src/package.json [hunk 9]
```diff
@@ -15,12 +15,16 @@
     "@mui/icons-material": "^5.10.6",
     "@mui/material": "^5.10.7",
     "@next-auth/firebase-adapter": "^1.0.1",
-    "firebase-admin": "^11.0.1",
+    "": "^9.16.0",
+    "firebase-admin": "^11.4.1",
+    "fs": "^0.0.1-security",
```
[the remark this answers is not in the document]
Comment (johnnymcmike) [reply src/package.json:20]: No idea, went through a period of thinking some bugs were package-based (fs in particular only works on server side and i was trying to call it indirectly on client) and typed `npm i fs`. Don't have an explanation for the blank entry. Changing this stuff shouldn't break what I have working.

[the remark this answers is not in the document]
Comment (teamrocket77) [reply src/package.json:20]: I was having some issues with the `fs` (supposedly not installed) modules when troubleshooting the env issue. I think it is installed by default, not sure tho

File: src/pages/api/applicants/count/bymonth.js (hunk 1 above)
Comment (teamrocket77) [reply src/pages/api/applicants/count/bymonth.js:1]: ^^ Agreed, dynamic is the way should do nicely for any changes we or anyone else may want to make., Maybe this should be just an outline since there could be lots of functionality implied here. && it's already a large PR

File: src/pages/logintest.js (hunk 8 above)
Comment (teamrocket77) [reply src/pages/logintest.js:1]: Can someone explain what's happening in this file? 
If it's a unit test we have them in the `src/cypress/e2e/*` folder, you can make a `backend` folder and prefix it with `backend.<something>.cy.js. There are examples in the front end folder.
We will get started writing something that runs the back end tests.
If you want a brief overview of cypress 
check [this out](https://github.com/KSU-Google-DSC/Webapp/blob/main/docs/automation/CYPRESS_README.md) 
Ping me with any issues, you don't have to write it by yourself, all for doing it together if you're down. However, I will have to have an understanding of what's happening here to write it

**Review by teamrocket77 — CHANGES_REQUESTED** [review]

File: src/pages/logintest.js (hunk 8 above)
Comment (johnnymcmike) [reply src/pages/logintest.js:1]: It's not a unit test, it's a temporary login page that's not supposed to make it into prod, just for making sure auth actually works without having to rework the existing real login page myself. I copied it from the library's [example project here](https://github.com/gladly-team/next-firebase-auth/tree/v1.x/example). Sorry if the "test" name threw you off.

File: src/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (johnnymcmike) [reply src/package.json:20]: So in my research the "fs isn't installed" stuff apparently happens when you're trying to run code from a library meant for node.js in the browser. Since fs is the filestystem-accessing library and all. It turned out that my issue was with code and didn't have anything to do with fs not being installed or any packages being the wrong version.

File: src/initAuth.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    appPageURL: '/',
+    loginAPIEndpoint: '/api/auth/login',
+    logoutAPIEndpoint: '/api/auth/logout',
+    onLoginRequestError: (err) => {
+      console.error(err)
+    },
+    onLogoutRequestError: (err) => {
+      console.error(err)
+    },
+    firebaseAdminInitConfig: {
```
[the remark this answers is not in the document]
Comment (teamrocket77) [reply src/initAuth.js:15]: @johnnymcmike I don't know what the current state it but you might be able to use an optional. As seen [here](https://github.com/KSU-Google-DSC/Google-Summit-Solution/blob/api/signup/src/pages/signup.js)

File: src/database.js [hunk 11]
```diff
@@ -1,10 +1,18 @@
...
+let firestoreApp = !getApps().length ? initializeApp({
   credential: cert({
     projectId: process.env.FIREBASE_SERVICE_ACCOUNT_PROJECT_ID,
     privateKey: process.env.FIREBASE_SERVICE_ACCOUNT_PRIVATE_KEY ? process.env.FIREBASE_SERVICE_ACCOUNT_PRIVATE_KEY.replace(/\\n/g, '\n') : undefined,
     clientEmail: process.env.FIREBASE_SERVICE_ACCOUNT_CLIENT_EMAIL,
   }),
   databaseURL: process.env.FIRESTORE_SERVICE_DATABASE_URL
-}, "COMMON") : getApp("COMMON");
\ No newline at end of file
+}) : getApp();
```
Comment (teamrocket77) [inline src/database.js:16]: Why was the `COMMON` string removed?

File: src/pages/logintest.js (hunk 8 above)
Comment (teamrocket77) [reply src/pages/logintest.js:1]: @johnnymcmike if it shouldn't make it into prod why do we have it?

File: src/pages/api/applicants/submit.js [hunk 12]
```diff
@@ -1,15 +1,21 @@
...
-    }
-    const namekey = `${data.firstName}-${data.lastName}-${data.email}`.toLowerCase();
-    const newref = Firestore_Database.collection('applicants').doc(namekey);
-    await newref.set(data);
-    return true;
+import { Firestore_Database } from '../../../database';
+import { getUserFromCookies } from 'next-firebase-auth';
+
+export async function submitApplicant(data, authCookie) {
+  if (!(data.firstName && data.lastName && data.email)) {//TODO: move this validation to an adapter?
```
Comment (teamrocket77) [inline src/pages/api/applicants/submit.js:5]: I see that you have a question here. Isn't it possible to do on the frontend? Wouldn't backend processing and validation be slow. [Validation from Moz](https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation#using_built-in_form_validation)

File: src/pages/api/applicants/submit.js [hunk 13]
```diff
@@ -1,15 +1,21 @@
...
+    return false;
+  }
+  if (!(typeof data.firstName === 'string' && typeof data.lastName === 'string' && typeof data.email === 'string')) {//TODO: move this validation to an adapter?
+    return false;
+  }
+  if ((data.firstName + data.lastName + data.email).contains('<script')) {//one more XSS check to be safe
```
Comment (teamrocket77) [inline src/pages/api/applicants/submit.js:11]: Nothing wrong with additional protection but in our `next.config.js` I think we filter for this already

File: src/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (teamrocket77) [reply src/package.json:20]: @johnnymcmike so it'll be removed?

**Review by teamrocket77 — CHANGES_REQUESTED** [review]

File: src/database.js (hunk 11 above)
Comment (johnnymcmike) [reply src/database.js:16]: this was through trial and error - the library initializes a default firebase app, this code (on which we used to rely for initialization) initialized a firebase app with the name "COMMON", which was separate from the one the library was making. the goal was to have the library just init firebase-admin at the start and then be able to get it anywhere with firebase.getApp(). I remember this working for a time, but then I ended up having to bring back the whole code block in database.js because things would only get initialized if you visited an auth page, and it would give you an error if you started up the app and then visited a db page first.

TLDR: i'm not sure, but it works right now. if we need it to be named "COMMON" for some other reason, then even with that string restored it will probably break if a login page is visited before a database page.

File: src/pages/api/applicants/submit.js (hunk 13 above)
Comment (johnnymcmike) [reply src/pages/api/applicants/submit.js:11]: that's what sanjay said but it was suggested to add it anyway
